pkg/operator/controller/extension/extension/runtime/runtime.go

```diff
@@ -84,52 +90,106 @@ func (d *deployer) createOrUpdateResources(ctx context.Context, extension *opera
 		}
 	}
 
-	renderedChart, err := d.runtimeClientSet.ChartRenderer().RenderArchive(archive, extension.Name, d.gardenNamespace, utils.MergeMaps(helmValues, gardenerValues))
+	namespace := namespaceName(extension)
+	if err := d.ensureNamespace(ctx, namespace); err != nil {
+		return fmt.Errorf("failed ensuring namespace %q: %w", namespace, err)
+	}
+
+	renderedChart, err := d.runtimeClientSet.ChartRenderer().RenderArchive(archive, extension.Name, namespace, utils.MergeMaps(helmValues, gardenerValues))
 	if err != nil {
 		return fmt.Errorf("failed rendering Helm chart %q: %w", extension.Spec.Deployment.ExtensionDeployment.Helm.OCIRepository.GetURL(), err)
 	}
 
 	mrName := managedResourceName(extension)
-	if err := managedresources.CreateForSeed(ctx, d.runtimeClientSet.Client(), d.gardenNamespace, mrName, false, renderedChart.AsSecretData()); err != nil {
+	if err := managedresources.CreateForSeed(ctx, d.runtimeClientSet.Client(), namespace, mrName, false, renderedChart.AsSecretData()); err != nil {
 		return fmt.Errorf("failed creating ManagedResource: %w", err)
 	}
 
-	if err := managedresources.WaitUntilHealthyAndNotProgressing(ctx, d.runtimeClientSet.Client(), d.gardenNamespace, mrName); err != nil {
+	if err := managedresources.WaitUntilHealthyAndNotProgressing(ctx, d.runtimeClientSet.Client(), namespace, mrName); err != nil {
 		return fmt.Errorf("failed waiting for ManagedResource to be healthy: %w", err)
 	}
 	return nil
 }
 
 func (d *deployer) deleteResources(ctx context.Context, log logr.Logger, extension *operatorv1alpha1.Extension) error {
 	mrName := managedResourceName(extension)
+	namespace := namespaceName(extension)
 
-	log.Info("Deleting extension ManagedResource for runtime cluster if present", "managedResource", client.ObjectKey{Name: mrName, Namespace: d.gardenNamespace})
-	if err := managedresources.DeleteForSeed(ctx, d.runtimeClientSet.Client(), d.gardenNamespace, mrName); err != nil {
+	log.Info("Deleting extension ManagedResource for runtime cluster if present", "managedResource", client.ObjectKey{Name: mrName, Namespace: namespace})
+	if err := managedresources.DeleteForSeed(ctx, d.runtimeClientSet.Client(), namespace, mrName); err != nil {
 		return fmt.Errorf("failed deleting ManagedResource: %w", err)
 	}
 
-	if err := managedresources.WaitUntilDeleted(ctx, d.runtimeClientSet.Client(), d.gardenNamespace, mrName); err != nil {
+	if err := managedresources.WaitUntilDeleted(ctx, d.runtimeClientSet.Client(), namespace, mrName); err != nil {
 		return fmt.Errorf("failed waiting for ManagedResource to be deleted: %w", err)
 	}
+
+	if err := d.deleteNamespace(ctx, namespace); err != nil {
+		return fmt.Errorf("failed deleting namespace %q: %w", namespace, err)
+	}
+
```

Comment:
Shouldn't we wait and check if the namespace is really gone?

Reply:
Good point! Done